### DISCUSSION TITLE
GSoC 2022: Write-Ahead Log (WAL) infrastructure

### DISCUSSION THREAD

**Discussion by jinlang226** [pgmoneta/pgmoneta#55] [Ideas]
Hi all,

My name is Jinlang Wang and I recently graduated from the University of Pittsburgh. I'll go to graduate school at the University of Wisconsin-Madison this fall.

I will join the pgmoneta during GSoC 2022, and I will work on write-ahead log (WAL) infrastructure. 

This project aims to integrate the PostgreSQL Write-Ahead Log (WAL) data structures into pgmoneta.

I will work on this project by:
1. Study the PostgreSQL Write-Ahead Log (WAL) format
2. Replace usage of pg_receivewal inside pgmoneta with a pgmoneta native solution
3. Create a foundation for a pgmoneta library that can be used for WAL interaction

For more information, please refer to my [proposal](https://docs.google.com/document/d/1KKKDU6iP0GOkAMSdGRyxFJRgW964JFVROnpKkbzWyNw/edit#).

If you have any suggestions, please contact me! Thanks.

**Comment by jesperpedersen** [comment]
Welcome !

This is a big and complex project, but will be a corner stone for pgmoneta.

Studying and replacing pg_receivewal will create an understanding on what needs to be done in order to create the pgmoneta/WAL library.

I would start by using PostgreSQL 10 as a baseline for your work, and once you have a good understanding then explore the changes between the versions. As this is a clean-room implementation (e.g. no PostgreSQL includes) keeping the project in scope is very important; it is better to just support a few PostgreSQL versions with the library and have them be stable. Of course the scope is PostgreSQL 10 - 15, but we will dive deeper into the functionality once we get going.

The PostgreSQL source code is a good place to get started, but also resources like https://www.interdb.jp/pg/ can provide background.

Having a solution that can replace pg_receivewal for PostgreSQL 10 - 15 is def a great first step !

**Comment by jinlang226** [comment]
Hi Jesper,

After I run `./pgmoneta -c /home/jinlang/pgmoneta.conf `,

I run  `./pgmoneta-admin master-key`,

but I got the following error:

```
Could not write to master key file '/home/pgmoneta/.pgmoneta/master.key' due to No such file or directory
Error for master key
```


Then I tried make a `.pgmoneta` directory and `touch master-key` 

and then

```
[root@docker-desktop pgmoneta]# chmod 0700 /home/pgmoneta/.pgmoneta/master.key
[root@docker-desktop pgmoneta]# chown pgmoneta /home/pgmoneta
```

I got the following error

`[pgmoneta@docker-desktop src]$ ./pgmoneta-admin master-key
Wrong permissions for ~/.pgmoneta (must be 0700) `

How can I fix this? Thanks in advance!

**Reply by jesperpedersen** [reply]
The master key is created by

```
pgmoneta-admin master-key
```

See the GETTING_STARTED.md file

**Reply by jesperpedersen** [reply]
The directory should be created automatically for the current user

**Reply by jinlang226** [reply]
I run `./pgmoneta-admin master-key`,

but I got the following error:

```
Could not write to master key file '/home/pgmoneta/.pgmoneta/master.key' due to No such file or directory
Error for master key
```

**Reply by jesperpedersen** [reply]
And you run it as the `pgmoneta` user ?

**Reply by jesperpedersen** [reply]
Can you `gdb` your way through it and see where it fails ?

**Reply by jinlang226** [reply]
yes, I run it as pgmoneta user. I will try using gdb

**Comment by jinlang226** [comment]
Hi Jesper, 

I solved the previous problem. Since I run pgmoneta on docker, I used `chown` to give my user permission to write to the master key file. 

However, I encountered a new problem.

When I ` vi  /tmp/pgmoneta.log`, it shows:

``` 
[pgmoneta@docker-desktop jinlang]$ ls
2022-05-31 01:00:43 FATAL configuration.c:777 pgmoneta: unix_socket_dir is not a directory (/var/run/postgresql)
~
```

I have two docker containers. One is `pgmoneta`, and the other is `postgresql`. I checked the container of `postresql`, and it has the path of `/var/run/postgresql`, but `pgmoneta` does not. How can I link the container of `postgresql` to `pgmoneta` so that `pgmoneta` can find the path `/var/run/postgresql` in its local environment? 

Thanks.

**Reply by jesperpedersen** [reply]
So, it was just a permission problem ?

You don't need a `/var/run/postgresql/` directory on the pgmoneta node - just pick a directory like `/tmp` for the Unix Domain Socket -- will show up as `/tmp/.s.pgmoneta` then.

See https://github.com/pgmoneta/pgmoneta/blob/main/doc/tutorial/01_install.md or https://github.com/jesperpedersen/podman-pgsql/tree/master/pgsql14-pgmoneta-rocky8 for more hints on initial installation

**Reply by jinlang226** [reply]
Yes, it was a permission problem.

**Comment by jinlang226** [comment]
Hi Jesper,

I have some ideas regarding the WAL project.

Firstly, I printed out the cmd of `pgmoneta_wal`, and found it looks like this:

```
PGPASSWORD="secretpassword" /usr/pgsql-14/bin//pg_receivewal -h localhost -p 5432 -U repl --no-loop --no-password -D /home/kk/GSoC/upload/pgmoneta-plus/backup/primary/wal/
```

It looks like there is a local Postgres database. The program calls `pg_receivewal`, and start a socket connection. To replace the original wal method, I want to reuse the socket connection in the above cmd. That is, replicate the remote directory to local, so we do not need a local postgres database. Do you have any suggestions? 

Secondly, in `pgmoneta.h`, 

```
struct server
{
   char name[MISC_LENGTH];             /**< The name of the server */
   char host[MISC_LENGTH];             /**< The host name of the server */
   int port;                           /**< The port of the server */
   char username[MAX_USERNAME_LENGTH]; /**< The user name */
   char backup_slot[MISC_LENGTH];      /**< The backup slot name */
   char wal_slot[MISC_LENGTH];         /**< The WAL slot name */
   char follow[MISC_LENGTH];           /**< Follow a server */
   int retention;                      /**< The retention for the server */
   bool synchronous;                   /**< Run in synchronous mode */
   atomic_bool backup;                 /**< Is there an active backup */
   atomic_bool delete;                 /**< Is there an active delete */
   int wal_size;                       /**< The size of the WAL files */
   bool wal_streaming;                 /**< Is WAL streaming active */
   bool valid;                         /**< Is the server valid */
} __attribute__ ((aligned (64)));
```

Can you elaborate on what does `wal_streaming` (what is  wal streaming) and ` char follow[MISC_LENGTH];           /**< Follow a server */` mean?

Thanks!

**Reply by jesperpedersen** [reply]
The command string is based on your configuration (`PGPASSWORD`, `localhost`, `5432` and `repl`), and the rest is for the `pg_receivewal` application. So, you will use the same in your solution - using the `struct user` and `struct server` to connect to the server - you will use `pgmoneta_server_authenticate` for that. Then it is a matter of creating the WAL stream. You can setup tools like https://github.com/jesperpedersen/pgprtdbg/ to look at what flowing on the wire.

`wal_streaming` is a variable that shows if WAL is being streamed from a server and then used for creating warnings in the log.

`follow` is for following another server - if `[primary]` fails (it crashed), then `[secondary]` will take over if it has a `follow=primary` in its configuration.

**Reply by jinlang226** [reply]
I noticed in the code of postgresql, `receivelog.c` used `PQexec(conn, query)` to connect to postgresql and get the query (https://github.com/postgres/postgres/blob/master/src/bin/pg_basebackup/receivelog.c#L545). This method is from `libpq-fe.h`.

Should I use the method you mentioned above to connect to postgresql database, and then use methods from `libpq-fe.h` to get the query? I think they might do the same thing.

Thanks

**Reply by jesperpedersen** [reply]
You have to make a clean-room implementation so you will depend on code only in our repository.

After the call to `pgmoneta_server_authenticate` you will have an authenticated socket connection to PostgreSQL. Using that socket you can then create messages (`message.h|c`) and communicate with the server. So, you need to figure out which messages that you need and how that translate into a similar functionality that we have using `pg_receiverwal` now.

We should be using the streaming protocol so https://www.postgresql.org/docs/current/protocol-replication.html provides a starting point.

**Reply by jinlang226** [reply]
Since the implementation of `pg_receivewal` in postgres heavily depends on the `libpq` library, so I used some api from `libpq` to replace the `pg_receivewal` call in pgmoneta. I forked the repo and this is the new branch https://github.com/jinlang226/pgmoneta/tree/wal

My next step will be replacing `PQconnectdb` (from libpq) with `pgmoneta_server_authenticate` since the fundamental implementation is socket.  Also, I will try to replace the communication with `message.h|c`, and try to find a way to acknowledge reply. 

Do you have any suggestions? Thanks in advance!

**Reply by jesperpedersen** [reply]
Yeah, you can start that way in order to get a deeper understanding of the area
